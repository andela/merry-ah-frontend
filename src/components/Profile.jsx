import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Route, Switch } from 'react-router-dom';
import ProfileEdit from './ProfileEdit.jsx';
import {
  getUserProfile,
  getUserArts,
  editProfile,
} from '../actions/profileActions';
import SiderBar from './SideBar.jsx';
import ProfileBanner from './ProfileBanner.jsx';
import ArtList from './ArtList.jsx';
/**
 * Profile Componenet
 */
export class Profile extends Component {
  state = {
    bio: '',
    image: '',
    userType: '',
  };

  /**
 * @param {function} componentDidMount
 *  @returns {jsx} jsx
 */
  componentDidMount() {
    this.props.getUserProfile(this.props.auth.user.id);
    this.props.getUserArts(this.props.auth.user.id);
  }

  /**
   *  @param {number} event
 *  @returns {jsx} jsx
 */
  handleEditProfile = (event) => {
    const {
      bio,
      imgURL,
    } = this.props.user.userProfile.profile;
    const {
      userType,
    } = this.props.user.userProfile;
    event.preventDefault();
    const payload = {
      userType: this.state.userType || userType,
      bio: this.state.bio || bio,
      imgURL: this.state.image || imgURL,
    };
    this.props.editProfile(payload);
  }

  /**
 * @param {number} event
 *  @returns {jsx} jsx
 */
  handleChange = (event) => {
    if (event.target.id === 'image') {
      this.setState({
        image: event.target.files[0],
      });
    } else {
      this.setState({
        [event.target.id]: event.target.value,
      });
    }
  }

  /**
   * @param {function} render
   *  @returns {jsx} jsx
   */
  render() {
    const { isGettingArts } = this.props.user;
    return (

      <div className='row'>

        <div className='col-md-2 p-0'>
          <SiderBar />
        </div>

        <div className='col-md-10 p-0'>
          <ProfileBanner
            user={this.props.user}
          />
          <Switch>
            <Route
              path='/profile/edit-profile'
              render={props => (
                <ProfileEdit
                  user={this.props.user}
                  handleEditProfile={this.handleEditProfile}
                  handleChange={this.handleChange}
                  {...props}
                />
              )}
            />
            <Route
              exact
              path='/profile'
              render={state => (
                <ArtList
                  user={this.props.user}
                  isGettingArts={isGettingArts}
                  {...state}
                />
              )}
            />
          </Switch>
        </div>

      </div>
    );
  }
}

Profile.propTypes = {
  auth: PropTypes.shape({
    user: PropTypes.shape({
      id: PropTypes.number.isRequired,
    }),
  }),
  user: PropTypes.shape({
    isGettingArts: PropTypes.bool.isRequired,
    userProfile: PropTypes.shape({
      userType: PropTypes.string,
      profile: PropTypes.shape({
        bio: PropTypes.string.isRequired,
        imgURL: PropTypes.string.isRequired,
      }),
    }),
  }),
  getUserProfile: PropTypes.func.isRequired,
  editProfile: PropTypes.func.isRequired,
  getUserArts: PropTypes.func.isRequired,
};
export const mapStateToProps = state => ({
  auth: state.auth,
  user: state.user,
});

export default connect(mapStateToProps,
  { getUserProfile, getUserArts, editProfile })(Profile);
