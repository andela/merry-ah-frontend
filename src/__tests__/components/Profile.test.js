/* eslint-disable import/no-unresolved */
import 'babel-polyfill';
import React from 'react';
import { shallow } from 'enzyme';
import Profile from '../../components/Profile.jsx';
import PrivateRoute from '../../components/PrivateRoute.jsx';

let wrapper;
beforeEach(() => {
  wrapper = shallow(<PrivateRoute component={Profile} />);
});
describe('Profile Component', () => {
  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
