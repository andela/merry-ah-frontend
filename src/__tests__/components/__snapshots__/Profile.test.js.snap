// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileBanner Component should match snapshot 1`] = `
<MemoryRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "phq0ko",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "phq0ko",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Profile
      auth={
        Object {
          "user": Object {
            "id": 7,
          },
        }
      }
      editProfile={[Function]}
      getUserArts={[Function]}
      getUserProfile={[Function]}
      user={
        Object {
          "isGettingArts": false,
          "userArts": Array [
            Object {
              "Category": Object {
                "categoryName": "Poetry",
              },
              "description": "some descriptions",
              "featuredImg": "image.url",
              "slug": "isuguixu",
              "title": "some title",
            },
          ],
          "userProfile": Object {
            "id": 7,
            "profile": Object {
              "bio": "my bio",
              "firstName": "Juliet",
              "imgURL": "img.URL",
              "lastName": "Ezekwe",
            },
            "userType": "artist",
          },
        }
      }
    >
      <div
        className="row"
      >
        <div
          className="col-md-2 p-0"
        >
          <SideBar>
            <div
              className="wrapper"
            >
              <nav
                id="sidebar"
              >
                <ul
                  className="list-unstyled components"
                >
                  <li
                    className="active"
                  >
                    <a
                      className="side"
                    >
                      <i
                        className="fa fa-user"
                      />
                       My Page
                    </a>
                  </li>
                  <li>
                    <a
                      className="side"
                    >
                      <i
                        className="fa fa-cog"
                      />
                       Settings
                    </a>
                  </li>
                </ul>
              </nav>
              <div
                className="clear-fix"
              />
            </div>
          </SideBar>
        </div>
        <div
          className="col-md-10 p-0"
        >
          <ProfileBanner
            user={
              Object {
                "isGettingArts": false,
                "userArts": Array [
                  Object {
                    "Category": Object {
                      "categoryName": "Poetry",
                    },
                    "description": "some descriptions",
                    "featuredImg": "image.url",
                    "slug": "isuguixu",
                    "title": "some title",
                  },
                ],
                "userProfile": Object {
                  "id": 7,
                  "profile": Object {
                    "bio": "my bio",
                    "firstName": "Juliet",
                    "imgURL": "img.URL",
                    "lastName": "Ezekwe",
                  },
                  "userType": "artist",
                },
              }
            }
          >
            <div
              className="col-md-12 p-0"
            >
              <div
                className="profile-banner"
              >
                <div
                  id="profile-image"
                >
                  <div
                    style={
                      Object {
                        "backgroundImage": "url(img.URL)",
                        "backgroundPosition": "center center",
                        "backgroundSize": "cover",
                        "borderRadius": "50%",
                        "height": "215px",
                        "left": "2em",
                        "position": "absolute",
                        "width": "215px",
                      }
                    }
                  >
                    <h2
                      style={
                        Object {
                          "margin": "2.5em 0px 0px 7em",
                        }
                      }
                    >
                      Juliet
                      Ezekwe
                       
                    </h2>
                  </div>
                </div>
                <Link
                  replace={false}
                  to="/profile/edit-profile"
                >
                  <a
                    href="/profile/edit-profile"
                    onClick={[Function]}
                  >
                    <button
                      className="btn btn-primary btn-lg"
                      id="edit-button"
                      style={
                        Object {
                          "cursor": "pointer",
                        }
                      }
                    >
                      Edit Profile
                    </button>
                  </a>
                </Link>
              </div>
              <div
                className="row"
              >
                <div
                  className="col-md-8"
                >
                  Â 
                </div>
                <div
                  className="col-md-4"
                >
                  <div
                    id="aside"
                  >
                    <div
                      className="row"
                    >
                      <div
                        className="col-3 aside"
                      >
                         
                        <strong>
                          1
                        </strong>
                        <i
                          className="aside-text"
                        >
                           Artworks
                        </i>
                         
                      </div>
                      <div
                        className="col-3 aside"
                      >
                         
                        <strong>
                          253
                        </strong>
                        <i
                          className="aside-text"
                        >
                           Followers
                        </i>
                      </div>
                      <div
                        className="col-3 aside"
                      >
                         
                        <strong>
                          93
                        </strong>
                        <i
                          className="aside-text"
                        >
                           Following
                        </i>
                      </div>
                      <div
                        className="col-3 aside"
                      >
                         
                        <strong>
                          5
                        </strong>
                        <br />
                        <i
                          className="aside-text"
                        >
                           Sold
                        </i>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </ProfileBanner>
          <Switch />
        </div>
      </div>
    </Profile>
  </Router>
</MemoryRouter>
`;
